import os
from abc import ABCMeta
from pathlib import Path
from typing import Callable, Optional

from PyQt5.QtCore import QObject, Qt
from PyQt5.QtGui import QFont, QKeySequence, QMovie, QPixmap
from PyQt5.QtWidgets import (
    QApplication,
    QLabel,
    QLineEdit,
    QPushButton,
    QSizePolicy,
    QWidget,
)

from src.display.base_display import BaseDisplay
from src.utils.resource_finder import find_assets_dir


# åˆ›å»ºå…¼å®¹çš„å…ƒç±»
class CombinedMeta(type(QObject), ABCMeta):
    pass


class GuiDisplay(BaseDisplay, QObject, metaclass=CombinedMeta):
    def __init__(self):
        super().__init__()
        QObject.__init__(self)
        self.app = None
        self.root = None

        # UIæ§ä»¶
        self.status_label = None
        self.emotion_label = None
        self.tts_text_label = None
        self.manual_btn = None
        self.abort_btn = None
        self.auto_btn = None
        self.mode_btn = None
        self.text_input = None
        self.send_btn = None

        # è¡¨æƒ…ç®¡ç†
        self.emotion_movie = None
        self._emotion_cache = {}
        self._last_emotion_name = None

        # çŠ¶æ€ç®¡ç†
        self.auto_mode = False
        self._running = True
        self.current_status = ""
        self.is_connected = True

        # å›è°ƒå‡½æ•°
        self.button_press_callback = None
        self.button_release_callback = None
        self.mode_callback = None
        self.auto_callback = None
        self.abort_callback = None
        self.send_text_callback = None

        # ç³»ç»Ÿæ‰˜ç›˜ç»„ä»¶
        self.system_tray = None

    async def set_callbacks(
        self,
        press_callback: Optional[Callable] = None,
        release_callback: Optional[Callable] = None,
        mode_callback: Optional[Callable] = None,
        auto_callback: Optional[Callable] = None,
        abort_callback: Optional[Callable] = None,
        send_text_callback: Optional[Callable] = None,
    ):
        """
        è®¾ç½®å›è°ƒå‡½æ•°.
        """
        self.button_press_callback = press_callback
        self.button_release_callback = release_callback
        self.mode_callback = mode_callback
        self.auto_callback = auto_callback
        self.abort_callback = abort_callback
        self.send_text_callback = send_text_callback

        # ä¸å†æ³¨å†ŒçŠ¶æ€ç›‘å¬å›è°ƒï¼Œç”±update_statusç›´æ¥å¤„ç†æ‰€æœ‰é€»è¾‘

    def _on_manual_button_press(self):
        """
        æ‰‹åŠ¨æ¨¡å¼æŒ‰é’®æŒ‰ä¸‹äº‹ä»¶å¤„ç†.
        """
        if self.manual_btn and self.manual_btn.isVisible():
            self.manual_btn.setText("æ¾å¼€ä»¥åœæ­¢")
        if self.button_press_callback:
            self.button_press_callback()

    def _on_manual_button_release(self):
        """
        æ‰‹åŠ¨æ¨¡å¼æŒ‰é’®é‡Šæ”¾äº‹ä»¶å¤„ç†.
        """
        if self.manual_btn and self.manual_btn.isVisible():
            self.manual_btn.setText("æŒ‰ä½åè¯´è¯")
        if self.button_release_callback:
            self.button_release_callback()

    def _on_auto_button_click(self):
        """
        è‡ªåŠ¨æ¨¡å¼æŒ‰é’®ç‚¹å‡»äº‹ä»¶å¤„ç†.
        """
        if self.auto_callback:
            self.auto_callback()

    def _on_abort_button_click(self):
        """
        å¤„ç†ä¸­æ­¢æŒ‰é’®ç‚¹å‡»äº‹ä»¶.
        """
        if self.abort_callback:
            self.abort_callback()

    def _on_mode_button_click(self):
        """
        å¯¹è¯æ¨¡å¼åˆ‡æ¢æŒ‰é’®ç‚¹å‡»äº‹ä»¶.
        """
        if self.mode_callback:
            if not self.mode_callback():
                return

        self.auto_mode = not self.auto_mode

        if self.auto_mode:
            self._update_mode_button_status("è‡ªåŠ¨å¯¹è¯")
            self._switch_to_auto_mode()
        else:
            self._update_mode_button_status("æ‰‹åŠ¨å¯¹è¯")
            self._switch_to_manual_mode()

    def _switch_to_auto_mode(self):
        """
        åˆ‡æ¢åˆ°è‡ªåŠ¨æ¨¡å¼çš„UIæ›´æ–°.
        """
        if self.manual_btn and self.auto_btn:
            self.manual_btn.hide()
            self.auto_btn.show()

    def _switch_to_manual_mode(self):
        """
        åˆ‡æ¢åˆ°æ‰‹åŠ¨æ¨¡å¼çš„UIæ›´æ–°.
        """
        if self.manual_btn and self.auto_btn:
            self.auto_btn.hide()
            self.manual_btn.show()

    async def update_status(self, status: str, connected: bool):
        """
        æ›´æ–°çŠ¶æ€æ–‡æœ¬å¹¶å¤„ç†ç›¸å…³é€»è¾‘.
        """
        full_status_text = f"çŠ¶æ€: {status}"
        self._safe_update_label(self.status_label, full_status_text)

        # æ—¢è·Ÿè¸ªçŠ¶æ€æ–‡æœ¬å˜åŒ–ï¼Œä¹Ÿè·Ÿè¸ªè¿æ¥çŠ¶æ€å˜åŒ–
        new_connected = bool(connected)
        status_changed = status != self.current_status
        connected_changed = new_connected != self.is_connected

        if status_changed:
            self.current_status = status
        if connected_changed:
            self.is_connected = new_connected

        # ä»»ä¸€å˜åŒ–éƒ½æ›´æ–°ç³»ç»Ÿæ‰˜ç›˜
        if status_changed or connected_changed:
            self._update_system_tray(status)

    async def update_text(self, text: str):
        """
        æ›´æ–°TTSæ–‡æœ¬.
        """
        self._safe_update_label(self.tts_text_label, text)

    async def update_emotion(self, emotion_name: str):
        """
        æ›´æ–°è¡¨æƒ…æ˜¾ç¤º.
        """
        if emotion_name == self._last_emotion_name:
            return

        self._last_emotion_name = emotion_name
        asset_path = self._get_emotion_asset_path(emotion_name)

        if self.emotion_label:
            try:
                self._set_emotion_asset(self.emotion_label, asset_path)
            except Exception as e:
                self.logger.error(f"è®¾ç½®è¡¨æƒ…GIFæ—¶å‘ç”Ÿé”™è¯¯: {str(e)}")

    def _get_emotion_asset_path(self, emotion_name: str) -> str:
        """
        è·å–è¡¨æƒ…èµ„æºæ–‡ä»¶è·¯å¾„ï¼Œè‡ªåŠ¨åŒ¹é…å¸¸è§åç¼€.
        """
        if emotion_name in self._emotion_cache:
            return self._emotion_cache[emotion_name]

        assets_dir = find_assets_dir()
        if not assets_dir:
            path = "ğŸ˜Š"
        else:
            emotion_dir = assets_dir / "emojis"
            # æ”¯æŒçš„åç¼€ä¼˜å…ˆçº§ï¼šgif > png > jpg > jpeg > webp
            candidates = [
                emotion_dir / f"{emotion_name}.gif",
                emotion_dir / f"{emotion_name}.png",
                emotion_dir / f"{emotion_name}.jpg",
                emotion_dir / f"{emotion_name}.jpeg",
                emotion_dir / f"{emotion_name}.webp",
            ]
            # ä¾æ¬¡åŒ¹é…
            found = next((p for p in candidates if p.exists()), None)

            # å…œåº•åˆ° neutral åŒæ ·è§„åˆ™
            if not found:
                neutral_candidates = [
                    emotion_dir / "neutral.gif",
                    emotion_dir / "neutral.png",
                    emotion_dir / "neutral.jpg",
                    emotion_dir / "neutral.jpeg",
                    emotion_dir / "neutral.webp",
                ]
                found = next((p for p in neutral_candidates if p.exists()), None)

            path = str(found) if found else "ğŸ˜Š"

        self._emotion_cache[emotion_name] = path
        return path

    def _set_emotion_asset(self, label, asset_path: str):
        """
        è®¾ç½®è¡¨æƒ…èµ„æºï¼ˆGIFåŠ¨å›¾æˆ–é™æ€å›¾ç‰‡ï¼‰ã€‚
        """
        if not label:
            return

        # å¦‚æœæ˜¯emojiå­—ç¬¦ä¸²ï¼Œç›´æ¥è®¾ç½®æ–‡æœ¬
        if not isinstance(asset_path, str) or "." not in asset_path:
            label.setText(asset_path or "ğŸ˜Š")
            return

        try:
            if asset_path.lower().endswith(".gif"):
                # GIF åŠ¨å›¾
                if hasattr(self, "_gif_movies") and asset_path in self._gif_movies:
                    movie = self._gif_movies[asset_path]
                else:
                    movie = QMovie(asset_path)
                    if not movie.isValid():
                        label.setText("ğŸ˜Š")
                        return
                    movie.setCacheMode(QMovie.CacheAll)
                    if not hasattr(self, "_gif_movies"):
                        self._gif_movies = {}
                    self._gif_movies[asset_path] = movie

                # å¦‚åˆ‡æ¢åˆ°æ–°çš„movieï¼Œåœæ­¢æ—§çš„ä»¥é¿å…CPUå ç”¨
                if getattr(self, "emotion_movie", None) is not None and self.emotion_movie is not movie:
                    try:
                        self.emotion_movie.stop()
                    except Exception:
                        pass

                self.emotion_movie = movie
                label.setSizePolicy(QSizePolicy.Expanding, QSizePolicy.Expanding)
                label.setAlignment(Qt.AlignCenter)
                label.setMovie(movie)
                movie.setSpeed(105)
                movie.start()
            else:
                # é™æ€å›¾ç‰‡ï¼šå¦‚æœ‰æ—§çš„GIFåœ¨æ’­æ”¾åˆ™åœæ­¢
                if getattr(self, "emotion_movie", None) is not None:
                    try:
                        self.emotion_movie.stop()
                    except Exception:
                        pass
                    self.emotion_movie = None

                pixmap = QPixmap(asset_path)
                if pixmap.isNull():
                    label.setText("ğŸ˜Š")
                    return
                label.setSizePolicy(QSizePolicy.Expanding, QSizePolicy.Expanding)
                label.setAlignment(Qt.AlignCenter)
                label.setPixmap(pixmap)

        except Exception as e:
            self.logger.error(f"è®¾ç½®GIFåŠ¨ç”»å¤±è´¥: {e}")
            label.setText("ğŸ˜Š")

    def _safe_update_label(self, label, text):
        """
        å®‰å…¨åœ°æ›´æ–°æ ‡ç­¾æ–‡æœ¬.
        """
        if label:
            try:
                label.setText(text)
            except RuntimeError as e:
                self.logger.error(f"æ›´æ–°æ ‡ç­¾å¤±è´¥: {e}")

    async def close(self):
        """
        å…³é—­çª—å£å¤„ç†.
        """
        self._running = False
        # åœæ­¢å¹¶æ¸…ç†GIFèµ„æºï¼Œé¿å…èµ„æºæ³„æ¼
        try:
            if getattr(self, "emotion_movie", None) is not None:
                try:
                    self.emotion_movie.stop()
                except Exception:
                    pass
                self.emotion_movie = None
            if hasattr(self, "_gif_movies") and isinstance(self._gif_movies, dict):
                for _m in list(self._gif_movies.values()):
                    try:
                        _m.stop()
                    except Exception:
                        pass
                self._gif_movies.clear()
        except Exception:
            pass
        if self.system_tray:
            self.system_tray.hide()
        if self.root:
            self.root.close()

    async def start(self):
        """
        å¯åŠ¨GUI.
        """
        try:
            # è®¾ç½®Qtç¯å¢ƒå˜é‡
            os.environ.setdefault("QT_LOGGING_RULES", "qt.qpa.fonts.debug=false")

            self.app = QApplication.instance()
            if self.app is None:
                raise RuntimeError("QApplicationæœªæ‰¾åˆ°ï¼Œè¯·ç¡®ä¿åœ¨qasyncç¯å¢ƒä¸­è¿è¡Œ")

            # å…³é—­æœ€åä¸€ä¸ªçª—å£è¢«å…³é—­æ—¶è‡ªåŠ¨é€€å‡ºåº”ç”¨çš„è¡Œä¸ºï¼Œç¡®ä¿æ‰˜ç›˜å¸¸é©»
            try:
                self.app.setQuitOnLastWindowClosed(False)
            except Exception:
                pass

            # å®‰è£…åº”ç”¨çº§äº‹ä»¶è¿‡æ»¤å™¨ï¼šæ”¯æŒç‚¹å‡»Dockå›¾æ ‡æ—¶æ¢å¤çª—å£
            try:
                self.app.installEventFilter(self)
            except Exception:
                pass

            # è®¾ç½®é»˜è®¤å­—ä½“
            default_font = QFont()
            default_font.setPointSize(12)
            self.app.setFont(default_font)

            # åŠ è½½UI
            from PyQt5 import uic

            self.root = QWidget()
            ui_path = Path(__file__).parent / "gui_display.ui"
            uic.loadUi(str(ui_path), self.root)

            # è·å–æ§ä»¶å¹¶è¿æ¥äº‹ä»¶
            self._init_ui_controls()
            self._connect_events()

            # åˆå§‹åŒ–ç³»ç»Ÿæ‰˜ç›˜
            self._setup_system_tray()

            # è®¾ç½®é»˜è®¤è¡¨æƒ…
            await self._set_default_emotion()

            # æ˜¾ç¤ºçª—å£
            self.root.show()

        except Exception as e:
            self.logger.error(f"GUIå¯åŠ¨å¤±è´¥: {e}", exc_info=True)
            raise

    def eventFilter(self, obj, event):
        """
        åº”ç”¨çº§äº‹ä»¶è¿‡æ»¤ï¼š
        - macOS ç‚¹å‡» Dock å›¾æ ‡ä¼šè§¦å‘ ApplicationActivate äº‹ä»¶
        - å½“ä¸»çª—å£å¤„äºéšè—/æœ€å°åŒ–æ—¶ï¼Œè‡ªåŠ¨æ¢å¤æ˜¾ç¤º
        """
        try:
            # å»¶è¿Ÿå¯¼å…¥ï¼Œé¿å…é¡¶å±‚å¾ªç¯ä¾èµ–
            from PyQt5.QtCore import QEvent
            from PyQt5.QtCore import Qt as _Qt

            if event and event.type() == QEvent.ApplicationActivate:
                if self.root and not self.root.isVisible():
                    self._show_main_window()
        except Exception as e:
            if hasattr(self, "logger"):
                self.logger.error(f"å¤„ç†åº”ç”¨æ¿€æ´»äº‹ä»¶å¤±è´¥: {e}")
        return False

    def _init_ui_controls(self):
        """
        åˆå§‹åŒ–UIæ§ä»¶.
        """
        self.status_label = self.root.findChild(QLabel, "status_label")
        self.emotion_label = self.root.findChild(QLabel, "emotion_label")
        self.tts_text_label = self.root.findChild(QLabel, "tts_text_label")
        self.manual_btn = self.root.findChild(QPushButton, "manual_btn")
        self.abort_btn = self.root.findChild(QPushButton, "abort_btn")
        self.auto_btn = self.root.findChild(QPushButton, "auto_btn")
        self.mode_btn = self.root.findChild(QPushButton, "mode_btn")
        self.settings_btn = self.root.findChild(QPushButton, "settings_btn")
        self.text_input = self.root.findChild(QLineEdit, "text_input")
        self.send_btn = self.root.findChild(QPushButton, "send_btn")

    def _connect_events(self):
        """
        è¿æ¥äº‹ä»¶.
        """
        if self.manual_btn:
            self.manual_btn.pressed.connect(self._on_manual_button_press)
            self.manual_btn.released.connect(self._on_manual_button_release)
        if self.abort_btn:
            self.abort_btn.clicked.connect(self._on_abort_button_click)
        if self.auto_btn:
            self.auto_btn.clicked.connect(self._on_auto_button_click)
            self.auto_btn.hide()
        if self.mode_btn:
            self.mode_btn.clicked.connect(self._on_mode_button_click)
        if self.text_input and self.send_btn:
            self.send_btn.clicked.connect(self._on_send_button_click)
            self.text_input.returnPressed.connect(self._on_send_button_click)
        if self.settings_btn:
            self.settings_btn.clicked.connect(self._on_settings_button_click)

        # è®¾ç½®çª—å£å…³é—­äº‹ä»¶
        self.root.closeEvent = self._closeEvent

        # å¿«æ·é”®ï¼šCtrl+, ä¸ Cmd+, æ‰“å¼€è®¾ç½®
        try:
            from PyQt5.QtWidgets import QShortcut
            QShortcut(QKeySequence("Ctrl+,"), self.root, activated=self._on_settings_button_click)
            QShortcut(QKeySequence("Meta+,"), self.root, activated=self._on_settings_button_click)
        except Exception:
            pass

    def _setup_system_tray(self):
        """
        è®¾ç½®ç³»ç»Ÿæ‰˜ç›˜.
        """
        try:
            # å…è®¸é€šè¿‡ç¯å¢ƒå˜é‡ç¦ç”¨ç³»ç»Ÿæ‰˜ç›˜ç”¨äºæ’éšœ
            if os.getenv("XIAOZHI_DISABLE_TRAY") == "1":
                self.logger.warning("å·²é€šè¿‡ç¯å¢ƒå˜é‡ç¦ç”¨ç³»ç»Ÿæ‰˜ç›˜ (XIAOZHI_DISABLE_TRAY=1)")
                return
            from src.views.components.system_tray import SystemTray

            self.system_tray = SystemTray(self.root)
            self.system_tray.show_window_requested.connect(self._show_main_window)
            self.system_tray.settings_requested.connect(self._on_settings_button_click)
            self.system_tray.quit_requested.connect(self._quit_application)

        except Exception as e:
            self.logger.error(f"åˆå§‹åŒ–ç³»ç»Ÿæ‰˜ç›˜ç»„ä»¶å¤±è´¥: {e}", exc_info=True)

    async def _set_default_emotion(self):
        """
        è®¾ç½®é»˜è®¤è¡¨æƒ….
        """
        try:
            await self.update_emotion("neutral")
        except Exception as e:
            self.logger.error(f"è®¾ç½®é»˜è®¤è¡¨æƒ…å¤±è´¥: {e}", exc_info=True)

    def _update_system_tray(self, status):
        """
        æ›´æ–°ç³»ç»Ÿæ‰˜ç›˜çŠ¶æ€.
        """
        if self.system_tray:
            self.system_tray.update_status(status, self.is_connected)

    def _show_main_window(self):
        """
        æ˜¾ç¤ºä¸»çª—å£.
        """
        if self.root:
            if self.root.isMinimized():
                self.root.showNormal()
            if not self.root.isVisible():
                self.root.show()
            self.root.activateWindow()
            self.root.raise_()

    def _quit_application(self):
        """
        é€€å‡ºåº”ç”¨ç¨‹åº.
        """
        self.logger.info("å¼€å§‹é€€å‡ºåº”ç”¨ç¨‹åº...")
        self._running = False

        if self.system_tray:
            self.system_tray.hide()

        try:
            from src.application import Application

            app = Application.get_instance()
            if app:
                # å¼‚æ­¥å¯åŠ¨å…³é—­æµç¨‹ï¼Œä½†è®¾ç½®è¶…æ—¶
                import asyncio

                from PyQt5.QtCore import QTimer

                loop = asyncio.get_event_loop()
                if loop.is_running():
                    # åˆ›å»ºå…³é—­ä»»åŠ¡ï¼Œä½†ä¸ç­‰å¾…
                    shutdown_task = asyncio.create_task(app.shutdown())

                    # è®¾ç½®è¶…æ—¶åå¼ºåˆ¶é€€å‡º
                    def force_quit():
                        if not shutdown_task.done():
                            self.logger.warning("å…³é—­è¶…æ—¶ï¼Œå¼ºåˆ¶é€€å‡º")
                            shutdown_task.cancel()
                        QApplication.quit()

                    # 3ç§’åå¼ºåˆ¶é€€å‡º
                    QTimer.singleShot(3000, force_quit)

                    # å½“shutdownå®Œæˆæ—¶æ­£å¸¸é€€å‡º
                    def on_shutdown_complete(task):
                        if not task.cancelled():
                            if task.exception():
                                self.logger.error(
                                    f"åº”ç”¨ç¨‹åºå…³é—­å¼‚å¸¸: {task.exception()}"
                                )
                            else:
                                self.logger.info("åº”ç”¨ç¨‹åºæ­£å¸¸å…³é—­")
                        QApplication.quit()

                    shutdown_task.add_done_callback(on_shutdown_complete)
                else:
                    # å¦‚æœäº‹ä»¶å¾ªç¯æœªè¿è¡Œï¼Œç›´æ¥é€€å‡º
                    QApplication.quit()
            else:
                QApplication.quit()

        except Exception as e:
            self.logger.error(f"å…³é—­åº”ç”¨ç¨‹åºå¤±è´¥: {e}")
            # å¼‚å¸¸æƒ…å†µä¸‹ç›´æ¥é€€å‡º
            QApplication.quit()

    def _closeEvent(self, event):
        """
        å¤„ç†çª—å£å…³é—­äº‹ä»¶.
        """
        # åªè¦ç³»ç»Ÿæ‰˜ç›˜å¯ç”¨ï¼Œå°±æœ€å°åŒ–åˆ°æ‰˜ç›˜
        if self.system_tray and (
            getattr(self.system_tray, "is_available", lambda: False)()
            or getattr(self.system_tray, "is_visible", lambda: False)()
        ):
            self.logger.info("å…³é—­çª—å£ï¼šæœ€å°åŒ–åˆ°æ‰˜ç›˜")
            # å»¶è¿Ÿéšè—ï¼Œé¿å…åœ¨closeEventä¸­ç›´æ¥æ“ä½œçª—å£å¼•å‘macOSå›¾å½¢æ ˆä¸ç¨³å®š
            try:
                from PyQt5.QtCore import QTimer
                QTimer.singleShot(0, self.root.hide)
            except Exception:
                try:
                    self.root.hide()
                except Exception:
                    pass
            # åœæ­¢GIFåŠ¨ç”»ï¼Œè§„é¿éšè—æ—¶çš„æ½œåœ¨å´©æºƒ
            try:
                if getattr(self, "emotion_movie", None) is not None:
                    self.emotion_movie.stop()
            except Exception:
                pass
            event.ignore()
        else:
            self._quit_application()
            event.accept()

    def _update_mode_button_status(self, text: str):
        """
        æ›´æ–°æ¨¡å¼æŒ‰é’®çŠ¶æ€.
        """
        if self.mode_btn:
            self.mode_btn.setText(text)

    async def update_button_status(self, text: str):
        """
        æ›´æ–°æŒ‰é’®çŠ¶æ€.
        """
        if self.auto_mode and self.auto_btn:
            self.auto_btn.setText(text)

    def _on_send_button_click(self):
        """
        å¤„ç†å‘é€æ–‡æœ¬æŒ‰é’®ç‚¹å‡»äº‹ä»¶.
        """
        if not self.text_input or not self.send_text_callback:
            return

        text = self.text_input.text().strip()
        if not text:
            return

        self.text_input.clear()

        try:
            import asyncio

            task = asyncio.create_task(self.send_text_callback(text))

            def _on_done(t):
                if not t.cancelled() and t.exception():
                    self.logger.error(f"å‘é€æ–‡æœ¬ä»»åŠ¡å¼‚å¸¸: {t.exception()}", exc_info=True)
            task.add_done_callback(_on_done)
        except Exception as e:
            self.logger.error(f"å‘é€æ–‡æœ¬æ—¶å‡ºé”™: {e}")

    def _on_settings_button_click(self):
        """
        å¤„ç†è®¾ç½®æŒ‰é’®ç‚¹å‡»äº‹ä»¶.
        """
        try:
            from src.views.settings import SettingsWindow

            settings_window = SettingsWindow(self.root)
            settings_window.exec_()

        except Exception as e:
            self.logger.error(f"æ‰“å¼€è®¾ç½®çª—å£å¤±è´¥: {e}", exc_info=True)

    async def toggle_mode(self):
        """
        åˆ‡æ¢æ¨¡å¼.
        """
        # è°ƒç”¨ç°æœ‰çš„æ¨¡å¼åˆ‡æ¢åŠŸèƒ½
        if hasattr(self, "mode_callback") and self.mode_callback:
            self._on_mode_button_click()
            self.logger.debug("é€šè¿‡å¿«æ·é”®åˆ‡æ¢äº†å¯¹è¯æ¨¡å¼")

    async def toggle_window_visibility(self):
        """
        åˆ‡æ¢çª—å£å¯è§æ€§.
        """
        if self.root:
            if self.root.isVisible():
                self.logger.debug("é€šè¿‡å¿«æ·é”®éšè—çª—å£")
                self.root.hide()
            else:
                self.logger.debug("é€šè¿‡å¿«æ·é”®æ˜¾ç¤ºçª—å£")
                self.root.show()
                self.root.activateWindow()
                self.root.raise_()
